core/build.rs

```diff
@@ -1,26 +1,230 @@
+            "cargo:warning=Building GraphQL binary for {}-{}...",
+            target_info.os, target_info.arch
+        );
+
+        build_graphql_binary(&graphql_dir, &final_exe_path);
+    } else {
+        println!("cargo:warning=GraphQL binary is up to date, skipping build");
+        // Still set the environment variable for the existing binary
+        println!("cargo:rustc-env=RINDEXER_GRAPHQL_EXE={}", final_exe_path.display());
+    }
+
+    // Register build dependencies
+    register_build_dependencies(&manifest_dir);
+}
 
-    if target_dir.exists() {
-        fs::remove_dir_all(&target_dir).expect("Failed to remove old resources directory");
+fn should_rebuild(exe_path: &Path, graphql_dir: &Path) -> bool {
+    // If binary doesn't exist, rebuild
+    if !exe_path.exists() {
+        println!("cargo:warning=Binary doesn't exist, rebuilding");
+        return true;
     }
-    fs::create_dir_all(&target_dir).expect("Failed to create resources directory");
-    for entry in fs::read_dir(resources_dir).expect("Failed to read resources directory") {
-        let entry = entry.expect("Failed to read directory entry");
-        let path = entry.path();
-        if path.is_file() {
-            fs::copy(&path, target_dir.join(path.file_name().expect("Failed to get file name")))
-                .expect("Failed to copy file");
+
+    // Check if package-lock.json is newer than the binary (dependencies changed)
+    let package_lock = graphql_dir.join("package-lock.json");
+    if package_lock.exists() {
+        if let (Ok(exe_time), Ok(lock_time)) = (
+            exe_path.metadata().and_then(|m| m.modified()),
+            package_lock.metadata().and_then(|m| m.modified())
+        ) {
+            if lock_time > exe_time {
+                println!("cargo:warning=package-lock.json is newer than binary, rebuilding");
+                return true;
+            }
+        }
+    }
+
+    // Check if any main source files are newer than the binary
+    let source_files = ["index.js", "package.json"];
+    if let Ok(exe_time) = exe_path.metadata().and_then(|m| m.modified()) {
+        for source_file in &source_files {
+            let source_path = graphql_dir.join(source_file);
+            if source_path.exists() {
+                if let Ok(source_time) = source_path.metadata().and_then(|m| m.modified()) {
+                    if source_time > exe_time {
+                        println!("cargo:warning=Source file {} is newer than binary, rebuilding", source_file);
+                        return true;
+                    }
+                }
+            }
+        }
+    }
+
+    false
+}
+
+struct TargetInfo {
+    os: String,
+    arch: String,
+    exe_name: String,
+}
+
+fn get_target_info() -> TargetInfo {
+    let os = env::var("CARGO_CFG_TARGET_OS").expect("CARGO_CFG_TARGET_OS not set");
+    let arch = env::var("CARGO_CFG_TARGET_ARCH").expect("CARGO_CFG_TARGET_ARCH not set");
+
+    let node_arch = match arch.as_str() {
```

Comment:
i think it may be worth having a github action which installs the binary and runs it and sees if it works? im not sure how else to test this on all operating systems and its a pretty risky merge

Reply:
we can add a smoke test right after the build

Reply:
yes i think it would be good just because i am in same boat only have a mac and the CI has everything so if we can have smoke tests of it actually being able to run it would be good - alongside worth checking https://github.com/joshstevens19/rindexer/blob/master/documentation/docs/public/install.sh this script now as i think if its in binary we can remove the resources part of it? https://github.com/joshstevens19/rindexer/blob/master/documentation/docs/public/install.sh#L80